bootloader/wscript

```diff
@@ -188,43 +192,64 @@ def set_lsb_compiler(ctx):
     ctx.env.append_value('LINKFLAGS', lsb_target_flag)
 
 
+def check_sizeof_pointer(ctx):
+
+    def check(type, expected):
+        # test code taken from autoconf resp. Scons: this is a pretty clever
+        # hack to find that a type is of a given size using only compilation.
+        # This speeds things up quite a bit compared to straightforward code
+        # actually running the code.
+        # Plus: This works cross :-)
+        fragment = '''
+        int main() {
+            static int test_array[1 - 2 * !(((long int) (sizeof(%s))) == %d)];
+            test_array[0] = 0;
+            return 0;
+        }''' % (type, expected)
+        return ctx.check_cc(fragment=fragment, execute=False, mandatory=False)
+
+
+    ctx.start_msg("Checking size of pointer")
+    for size in (4, 8):
+        if check("void *", size):
+            break
+    else:
+        ctx.end_msg(False)
+        ctx.fatal("Couldn't determine pointer size, only 32 or 64 bit are supported. Please use `--target-arch' to set the pointer size.")
+    ctx.end_msg(size)
+    return size
+
+
 @conf
-def detect_arch_cpu(ctx):
+def detect_arch(ctx):
     """
-    Handle options --target-arch and --target-cpu or use the same
-    architecture as the Python interpreter.
+    Handle options --target-arch or use the same
+    architecture as the compiler.
     """
-    # Get arch/cpu values either from CLI or detect it.
-    arch = ctx.options.target_arch or architecture()
-    # 'target_cpu is useful only on Windows for now.
-    cpu = ctx.options.target_cpu or ''  # Use empty string if not specified.
-
-    # Print message based on arch/cpu.
-    if ctx.options.target_arch or ctx.options.target_cpu:
-        if cpu == 'amd64':
-            ctx.env['MSVC_TARGETS'] = ['x64']
-            arch = '64bit'  # Force 64bit for amd64 cpu.
-        elif cpu == 'x86':
-            ctx.env['MSVC_TARGETS'] = ['x86']
-            arch = '32bit'  # Force 32bit for x86 cpu.
-            ctx.msg('Platform', 'Architecture manually chosen: x86')
-        ctx.msg('Platform', '%s-%s %s  manually chosen' % (platform.system(), arch, cpu))
+    try:
+        system = DESTOS_TO_SYSTEM[ctx.env.DEST_OS]
+    except KeyError:
+        ctx.fatal('Unrecognized target system: %s' % ctx.env.DEST_OS)
+
+    # Get arch values either from CLI or detect it.
+    if ctx.options.target_arch:
+        arch = ctx.options.target_arch
+        ctx.msg('Platform', '%s-%s manually chosen' % (system, arch))
     else:
-        ctx.msg('Platform', '%s-%s detected' % (platform.system(), arch))
+        # PyInstaller uses the result of platform.architecture() to determine
+        # the bits and this is testing the pointer size (via module struct).
+        # We do the same here.
+        arch = "%sbit" % (8 * check_sizeof_pointer(ctx))
```

Comment:
This will not ask the correct compiler though. On appveyor you're asking the 64 bit compiler whether 32 bit is available.